crates/bevy_app/src/plugin.rs

```diff
@@ -6,14 +6,22 @@ use std::any::Any;
 /// A collection of Bevy app logic and configuration.
 ///
 /// Plugins configure an [`App`]. When an [`App`] registers a plugin,
-/// the plugin's [`Plugin::build`] function is run.
+/// the plugin's [`Plugin::build`] function is run. By default, a plugin
+/// can only be added once to an [`App`]. If the plugin may need to be
+/// added twice or more, the function [`is_unique`](Plugin::is_unique)
+/// should be overriden to return `false`.
 pub trait Plugin: Downcast + Any + Send + Sync {
     /// Configures the [`App`] to which this plugin is added.
     fn build(&self, app: &mut App);
     /// Configures a name for the [`Plugin`] which is primarily used for debugging.
     fn name(&self) -> &str {
         std::any::type_name::<Self>()
     }
+    /// If the plugin can be meaningfully instantiated several times in an [`App`](crate::App),
+    /// override this method to return `false`.
+    fn is_unique(&self) -> bool {
+        true
```

Comment:
Not fully necessary in this PR, but potential for followup: extending the derive macro to allow specifying this whether a plugin is unique without an explicit impl would be very nice. Seems like something people might want fairly regularly.

Reply:
I don't think there's a derive macro for plugins yet?

Reply:
There's not yet; it's just a manual impl :)

Reply:
Oh shoot, brainfart. It's been a while since I've implemented one. Ignore this, haha.